presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/SqlInvokedFunctionNamespaceManagerConfig.java

```diff
@@ -60,17 +62,15 @@ public SqlInvokedFunctionNamespaceManagerConfig setFunctionInstanceCacheExpirati
         return this;
     }
 
-    public Set<RoutineCharacteristics.Language> getSupportedFunctionLanguages()
+    public Map<Language, FunctionImplementationType> getSupportedFunctionLanguages()
     {
         return supportedFunctionLanguages;
     }
 
     @Config("supported-function-languages")
     public SqlInvokedFunctionNamespaceManagerConfig setSupportedFunctionLanguages(String languages)
     {
-        ImmutableSet.Builder<RoutineCharacteristics.Language> languageBuilder = ImmutableSet.builder();
-        Splitter.on(',').omitEmptyStrings().trimResults().split(languages).forEach(language -> languageBuilder.add(new RoutineCharacteristics.Language(language)));
-        this.supportedFunctionLanguages = languageBuilder.build();
+        this.supportedFunctionLanguages = FUNCTION_LANGUAGES_CODEC.fromJson(languages);
```

Comment:
It seems a bit awkward to me that a json map needs to be specified in config, and also string-to-string json map is allowed. Since engine does need to understand the enums. What about changing the config properties to
```
remote-thrift-languages=language1,language2
```
This way, json is avoided, and user also don't need to specify the redundant entry for `Language=SQL`.
Even if we want to support implementation type `http`, the enum value needs to be added to spi and we can support that in FNM by adding a new config like
```
remote-http-languages=language3,language4
```

Reply:
Since it's a simple change, if we choose to do it, we might want to do it now to avoid changing the way config properties is specified in the future.

Reply:
Running SQL remotely is a valid potential option. It's a little strange to disallow it. The downside of your proposal is now we have to cross validate whether all languages has implementation type configured and all implementation type configuration is in the configured languages and a language doesn't have multiple implementation type configured... So cumbersome JSON doesn't sound like a terrible idea anymore in comparison. If this gets more annoying I think we should change this to a json file and have the json in a separate file (similar to file based resource groups configuration). Currently it's not annoying enough so I didn't bother. I can change to that if you prefer.

Reply:
The simplicity with JSON is due to the use of codec, which means we're not getting user-friendly error messages, but I don't have strong opinions here.

Reply:
> The simplicity with JSON is due to the use of codec, which means we're not getting user-friendly error messages, but I don't have strong opinions here.

Also that is's a map so we don't need to validate all keys have values or each value is only for 1 key. But ya, JSON codec is cool. ;)